translator/metrics/otlp_to_oc.go

```diff
@@ -0,0 +1,377 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"sort"
+
+	ocmetrics "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	otlpcommon "github.com/open-telemetry/opentelemetry-proto/gen/go/common/v1"
+	otlpmetrics "github.com/open-telemetry/opentelemetry-proto/gen/go/metrics/v1"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/internal"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	translatorcommon "github.com/open-telemetry/opentelemetry-collector/translator/common"
+)
+
+const (
+	invalidMetricDescriptorType = ocmetrics.MetricDescriptor_Type(-1)
+)
+
+// ResourceMetricsToMetricsData converts metrics from OTLP to internal (OpenCensus) format
+func ResourceMetricsToMetricsData(resourceMetrics *otlpmetrics.ResourceMetrics) consumerdata.MetricsData {
+
+	node, resource := translatorcommon.ResourceToOC(resourceMetrics.Resource)
+	md := consumerdata.MetricsData{
+		Node:     node,
+		Resource: resource,
+	}
+
+	if len(resourceMetrics.Metrics) == 0 {
+		return md
+	}
+
+	metrics := make([]*ocmetrics.Metric, 0, len(resourceMetrics.Metrics))
+	for _, metric := range resourceMetrics.Metrics {
+		metrics = append(metrics, metricToOC(metric))
+	}
+
+	md.Metrics = metrics
+	return md
+}
+
+func metricToOC(metric *otlpmetrics.Metric) *ocmetrics.Metric {
+	labelKeys := allLabelKeys(metric)
+	return &ocmetrics.Metric{
+		MetricDescriptor: descriptorToOC(metric.MetricDescriptor, labelKeys),
+		Timeseries:       dataPointsToTimeseries(metric, labelKeys),
+		Resource:         nil,
+	}
+}
+
+func descriptorToOC(descriptor *otlpmetrics.MetricDescriptor, labelKeys []*ocmetrics.LabelKey) *ocmetrics.MetricDescriptor {
+	if descriptor == nil {
+		return nil
+	}
+
+	return &ocmetrics.MetricDescriptor{
+		Name:        descriptor.Name,
+		Description: descriptor.Description,
+		Unit:        descriptor.Unit,
+		Type:        descriptorTypeToOC(descriptor.Type),
+		LabelKeys:   labelKeys,
+	}
+}
+
+func allLabelKeys(metric *otlpmetrics.Metric) []*ocmetrics.LabelKey {
+	// NOTE: OpenTelemetry and OpenCensus have different representations of labels:
+	// - OC has a single "global" ordered list of label keys per metric in the MetricDescriptor;
+	// then, every data point has an ordered list of label values matching the key index.
+	// - In OTLP, every label is represented independently via StringKeyValue,
+	// i.e. theoretically points in the same metric may have different set of labels.
+	//
+	// So what we do in this translator:
+	// - Scan all points and their labels to generate a unique set of all label keys
+	// used across the metric, sort them and set in MetricDescriptor.
```

Comment:
Sorting seems unnecessary.

Reply:
Agreed that is not mandatory, but there are 2 benefits:
- We can test the OpenCensus output labels without converting them to map (see `otlp_to_oc_test.go`
- We will always produce OC labels in the same order across multiple metrics, which might be preferred for further analysis / logging.

It does add extra overhead, so if you think it's unacceptable we can work around it in the tests.

Reply:
It is fine for now. We can benchmark later and change if needed. It is an implementation detail though, so let's not make it part of our public contract in any way (private tests relying on it is OK).